src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs

```diff
@@ -1495,6 +1497,161 @@ static bool extendedTypeIsOrDerivedFrom(NamedTypeSymbol hiddenContainer, NamedTy
+            {
+                current.MergePrioritized(additional);
+                return;
+            }
+
+            var currentSymbols = current.Symbols;
+            var additionalSymbols = additional.Symbols;
+
+            var currentCount = currentSymbols.Count;
+            var additionalCount = additionalSymbols.Count;
+            Debug.Assert(currentCount > 0);
+            Debug.Assert(additionalCount > 0);
+
+            for (int i = 0; i < additionalCount; i++)
+            {
+                var additionalSymbol = additionalSymbols[i];
+
+                removeLessSpecificNonMethodCurrentSymbols(additionalSymbol, currentSymbols, out bool isAdditionalSymbolHidden,
+                    currentCount, basesBeingResolved, ref useSiteInfo);
+
+                if (!isAdditionalSymbolHidden)
+                {
+                    currentSymbols.Add(additionalSymbol);
+                }
+            }
+
+            return;
+
+            // Determine whether we want to add the additional symbol
+            // and potentially remove some "hidden" non-method current symbols in the process
+            void removeLessSpecificNonMethodCurrentSymbols(Symbol additionalSymbol, ArrayBuilder<Symbol> currentSymbols, out bool isAdditionalSymbolHidden,
+                int currentCount, ConsList<TypeSymbol> basesBeingResolved, ref CompoundUseSiteInfo<AssemblySymbol> useSiteInfo)
+            {
+                isAdditionalSymbolHidden = false;
+                for (int j = 0; j < currentCount; j++)
+                {
+                    var currentSymbol = currentSymbols[j];
+                    if (currentSymbol is null)
+                    {
+                        continue;
+                    }
+
+                    if (currentSymbol.ContainingType.ExtendedTypeNoUseSiteDiagnostics is not { } currentContainerUnderlying
+                        || additionalSymbol.ContainingType.ExtendedTypeNoUseSiteDiagnostics is not { } additionalContainerUnderlying)
+                    {
+                        // We wouldn't have gathered these members if we didn't have an extended type for them.
+                        throw ExceptionUtilities.Unreachable();
+                    }
+
+                    if (additionalContainerUnderlying.Equals(currentContainerUnderlying, TypeCompareKind.ConsiderEverything))
+                    {
+                        // neither underlying type is more specific
+                    }
+                    else if (IsDerivedType(baseType: additionalContainerUnderlying, derivedType: currentContainerUnderlying,
+                        basesBeingResolved, this.Compilation, ref useSiteInfo))
+                    {
+                        // current symbol is more specific, it "hides" a less specific non-method/non-indexer additional symbol
+                        if (!IsMethodOrIndexer(additionalSymbol))
+                        {
+                            isAdditionalSymbolHidden = true;
```

Comment:
>isAdditionalSymbolHidden = true; [](http://example.com/codeflow?start=28&length=32)

Is there a good reason to continue after this fact is determined?

Reply:
We must continue as there may be other "current" symbols that need to be hidden.
Unlike previous `MergeHidingLookupResults` logic, we're not visiting extensions in any particular order relative to the hierarchy, so it's possible that the "current" bag already contains both less and more specific members.

Reply:
> We must continue as there may be other "current" symbols that need to be hidden.

I think the logic is hard to comprehend and, therefore, hard to get any confidence about its correctness. For example, if the additional symbol is hidden by a method in "derived type" why should it hide methods in "base type". I would expect us to collect all methods in this situation,  think this is what ```MergeHidingLookupResults``` would do.